otherlibs/systhreads/st_win32.h

```diff
@@ -158,30 +158,88 @@ Caml_inline void st_thread_yield(st_masterlock * m)
   EnterCriticalSection(m);
 }
 
+/* List of threads waiting on a mutex or on a condition variable.
+   Each thread is represented by its associated event. */
+
+struct st_wait_list {
+  HANDLE event;                  /* event of the first waiting thread */
+  struct st_wait_list * next;
+};
+
+/* Recover (or create) the event associated with the calling thread */
+static HANDLE st_event_for_current_thread(void)
+{
+  HANDLE ev = (HANDLE) TlsGetValue(st_thread_sem_key);
+  if (ev == 0) {
+    ev = CreateEvent(NULL,
+                     FALSE /*auto reset*/,
+                     FALSE /*initially unset*/,
+                     NULL);
+    if (ev == 0) return 0;
+    TlsSetValue(st_thread_sem_key, (void *) ev);
+  }
+  return ev;
+}
+
 /* Mutexes */
 
-typedef CRITICAL_SECTION * st_mutex;
+struct st_mutex_struct {
+  CRITICAL_SECTION lock;        /* protect the data structure */
+  enum { UNLOCKED, LOCKED } status;
+  struct st_wait_list * waiters; /* list of threads waiting to lock it */
+};
+
+typedef struct st_mutex_struct * st_mutex;
 
 static DWORD st_mutex_create(st_mutex * res)
 {
-  st_mutex m = caml_stat_alloc_noexc(sizeof(CRITICAL_SECTION));
+  st_mutex m = caml_stat_alloc_noexc(sizeof(struct st_mutex_struct));
   if (m == NULL) return ERROR_NOT_ENOUGH_MEMORY;
-  InitializeCriticalSection(m);
+  InitializeCriticalSection(&m->lock);
+  m->status = UNLOCKED;
+  m->waiters = NULL;
   *res = m;
   return 0;
 }
 
 static DWORD st_mutex_destroy(st_mutex m)
 {
-  DeleteCriticalSection(m);
+  DeleteCriticalSection(&m->lock);
   caml_stat_free(m);
   return 0;
 }
 
 Caml_inline DWORD st_mutex_lock(st_mutex m)
 {
+  DWORD rc;
+  HANDLE ev;
+  struct st_wait_list wait;
+
   TRACE1("st_mutex_lock", m);
-  EnterCriticalSection(m);
+  EnterCriticalSection(&m->lock);
+  while (m->status != UNLOCKED) {
+    ev = st_event_for_current_thread();
+    if (ev == NULL) { 
+      rc = GetLastError();
+      LeaveCriticalSection(&m->lock);
+      return rc;
+    }
+    /* Insert the current thread in the waiting list (atomically) */
+    wait.event = ev;
+    wait.next = m->waiters;
+    m->waiters = &wait;
+    LeaveCriticalSection(&m->lock);
+    /* Wait for our event to be signaled.  There is no risk of lost
+       wakeup, since we inserted ourselves on the waiting list of m
+       before releasing m's critical section */
+    TRACE1("st_mutex_lock: blocking on event", ev);
+    if (WaitForSingleObject(ev, INFINITE) == WAIT_FAILED)
+      return GetLastError();
+    TRACE1("st_mutex_lock: restarted", m);
+    EnterCriticalSection(&m->lock);
+  }
```

Comment:
```suggestion
  if (m->status != UNLOCKED)
    if ((ev = st_event_for_current_thread()) == NULL) {
      rc = GetLastError();
      LeaveCriticalSection(&m->lock);
      return rc;
    }
    while (m->status != UNLOCKED) {
      /* Insert the current thread in the waiting list (atomically) */
      wait.event = ev;
      wait.next = m->waiters;
      m->waiters = &wait;
      LeaveCriticalSection(&m->lock);
      /* Wait for our event to be signaled.  There is no risk of lost
         wakeup, since we inserted ourselves on the waiting list of m
         before releasing m's critical section */
      TRACE1("st_mutex_lock: blocking on event", ev);
      if (WaitForSingleObject(ev, INFINITE) == WAIT_FAILED)
        return GetLastError();
      TRACE1("st_mutex_lock: restarted", m);
      EnterCriticalSection(&m->lock);
    }
  }
```
(`ev` needs only to be initialised once)